src/cluster/replication.cc

```diff
@@ -874,6 +876,12 @@ Status ReplicationThread::fetchFile(int sock_fd, evbuffer *evbuf, const std::str
     UniqueEvbufReadln line(evbuf, EVBUFFER_EOL_CRLF_STRICT);
     if (!line) {
       if (auto s = util::EvbufferRead(evbuf, sock_fd, -1, ssl); !s) {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+          if (stop_flag_) {
+            return {Status::NotOK, "replication thread was stopped"};
+          }
+          continue;
+        }
```

Comment:
I think you can notice that to enter this block, the status must be OK (i.e. `if (auto s = ...; !s)`).

It means that, `evbuffer_read` must have a positive integer as a return value to indicate that the operation is successful. So I highly suspect that the errno will not be either of these two.

Reply:
Our QA team tested and it do resolve the issue (On X86-64 centos).

It depends on the return value of evbuffer_read, if it returns 0, it means the connection is EOF, or -1 if error occurred. So timeout will always return Status with NotOK and errno will be EAGAIN or EWOULDBLOCK as manual says:

if no data has been transferred and the timeout has been reached then -1 is returned with errno set to EAGAIN or EWOULDBLOCK, or EINPROGRESS (for [connect](https://linux.die.net/man/2/connect)(2)) just as if the socket was specified to be nonblocking.

Reply:
The other errors (not timeout), just returned as previous

Reply:
To enter this block, evbuffer_read should returns a non-positive integer to indicate the operation is not successful, so it only enters when the read operation fails.

If it fails, check the errno then we can retry the operation if timed out or quit if replication thread is stopped.

Reply:
Ahh you are right, sorry. It looks good to me now.